packages/hydrogen/src/cart/optimistic/useOptimisticCart.tsx

```diff
@@ -0,0 +1,128 @@
+import {useFetchers} from '@remix-run/react';
+import type {cartGetDefault} from '../queries/cartGetDefault';
+import {CartForm} from '../CartForm';
+import type {
+  Cart,
+  CartLine,
+  ProductVariant,
+} from '@shopify/hydrogen-react/storefront-api-types';
+import {
+  getOptimisticLineId,
+  isOptimisticLineId,
+} from './optimistic-cart.helper';
+
+export type OptimisticCart<T = ReturnType<typeof cartGetDefault>> = T & {
+  isOptimistic?: boolean;
+  lines: {nodes: Array<CartLine & {isOptimistic?: boolean}>};
+};
+
+/**
+ * @param cart The cart object from `context.cart.get()` returned by a server loader.
+ *
+ * @returns A new cart object augmented with optimistic state. Each cart line item that is optimistically added includes an `isOptimistic` property. Also if the cart has _any_ optimistic state, a root property `isOptimistic` will be set to `true`.
+ */
+export function useOptimisticCart<
+  DefaultCart = ReturnType<typeof cartGetDefault>,
+>(cart: DefaultCart): OptimisticCart<DefaultCart> {
+  const fetchers = useFetchers();
+
+  if (!fetchers || !fetchers.length) return cart as OptimisticCart<DefaultCart>;
+
+  let isOptimistic = false;
+
+  const optimisticCart = cart
```

Comment:
I noticed quite a few re-renders here. Should we maybe memoize this structuredClone that could be performance intensive on larger carts? Maybe we should add `updatedAt` to the default cart query and use that to trigger the `useMemo` refresh

```
  const optimisticCart = useMemo(() => cart
    ? (structuredClone(cart) as OptimisticCart<DefaultCart>)
    : ({lines: {nodes: []}} as OptimisticCart<DefaultCart>)
  , [cart.updatedAt]);
```